### PR TITLE
feat: Modern trie code with better memory management

### PR DESCRIPTION
**Pull request by Anmol3299**
#### Description of Change
The older implementation has very big memory leaks as there is no delete operation for the raw pointers used and assigned. The newer implementation uses modern c++ code and solves the memory leak issue too. I have tried to document the code as well as I could have.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTION.md#New-File-Name-guidelines)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTION.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] Sort by alphabetical order
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: Modernize code and fix memory leaks of trie data structure.

### CODE REVIEW

File: data_structure/trie_modern.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
+#include <iostream>
```
Comment (kvedala) [inline data_structure/trie_modern.cpp:1]: Please add the comment structure at the Beijing of the file. This enables the document generating tool to understand and read the file. refer -> https://github.com/TheAlgorithms/C-Plus-Plus/blob/16fec03ac899cdc847f20e553de7850a83caaa71/search/binary_search.h

File: data_structure/trie_modern.cpp [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+#include <memory>  // for std::shared_ptr<>
+#include <string>  // for std::string class
```
Comment (kvedala) [inline data_structure/trie_modern.cpp:3]: the `<string>` header is specific to Microsoft Visual C++. To make this generic, please use the following:
```c
#ifdef _MSC_VER
#include <string>
#else
#include <cstring>
#endif
```

File: data_structure/trie_modern.cpp [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+            curr = curr->character[index];
+        }
+        curr->isEndOfWord = true;
+    }
+
+    /**
+     * Search if a word is present in trie or not.
+     * @param word which is needed to be searched in the trie.
+     * @return if the word is found in the trie and isEndOfWord is set to true,
```
Comment (kvedala) [inline data_structure/trie_modern.cpp:121]: for return statements, you can also break then into two as such:
```
* @return True if word is found and ...
* @return False if word is not found and ...
```

File: data_structure/trie_modern.cpp [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    // Function to remove the word which calls the helper function.
+    void removeWord(const std::string& word) {
+        m_root = removeWordHelper(word, m_root, 0);
+    }
+
+ private:
+    std::shared_ptr<TrieNode> m_root;
+};
+
+int main() {
```
Comment (kvedala) [inline data_structure/trie_modern.cpp:148]: For the main, you can use a simple comment block like:
```
/**
 * Main function
 */
```

Comment (Anmol3299) [reply data_structure/trie_modern.cpp:159]: Pushed with all the suggestions

**Pull request merged**